### ISSUE TITLE
Average impact 

### ISSUE THREAD

**Issue by kdzwinel** [patrickhulce/third-party-web#39]
Hey 👋 Thank you again for this insightful project!

I've been contemplating usefulness of the 'average impact' metric. As far as I understand from readme, origin's impact is not aggregated at the website level e.g. if are two different scripts from `ads.example` running on `news.example` and each of them takes `50ms` of total CPU time, average is calculated as `50 + 50 / 2 requests` (`=50ms`) and not `50 + 50 / 1 website` (`=100ms`). Because of that, current method is rewarding origins that split their work between multiple scripts which doesn't seem fair to me. Grouping execution time per origin (or better yet, tld or entity) at the website level would significantly change order of entities in the "Third Parties by Category" section [1].

WDYT about such change? Did I miss something in my deliberations? Is such grouping feasible with current dataset? Or would it make more sense to take a step back and create a new Lighthouse audit that groups total CPU time by entity (perhaps using entity data from this project)?

[1] Average number of scripts per website for chosen, high prevalence, origins (those numbers come form my own small crawl):

script.hotjar.com - 1 (avg per resource - 19ms, avg per site - 19ms)
code.jquery.com - 1.14
www.googletagservices.com - 2.67
s7.addthis.com - 3.0
platform.twitter.com - 3.2
www.youtube.com - 3.38
t.sharethis.com - 5.05
www.facebook.com - 5.77  (avg per resource - 10.9ms, avg per site - 63ms)

**Comment by patrickhulce** [comment]
Definitely, thanks for filing @kdzwinel I've been trying to think about an efficient way to restructure the queries to address this problem myself!

I don't think grouping per entity is really feasible with a SQL query (other than maybe for the top-n entities? But maybe it's possible to have custom functions with ~100K of data in them? 😄 ) which puts us into this limbo situation where it'd be half per-script, half per-origin. The only way I see to truly address this would be going with some custom processing where we can use the entity data in its entirety.

Summoning @rviscomi for his thoughts. Any way I can explore Rick that would let me run some JS that has all the entity data baked in across the HTTPArchive lighthouse dataset?

**Comment by rviscomi** [comment]
Happy to help but not sure I understand the problem statement. Do you have a sample query or something that you could share that demonstrates what you're trying to do?

**Comment by patrickhulce** [comment]
@rviscomi the problem is that we are treating every individual *script* as a data point for the origin `GROUP BY` instead of every *page* as a data point for the `GROUP BY`.

This comes down to the problem that in order to group all of the impact for an entity per page, we essentially need to run the `require('third-party-web').getEntity` function on each entry in the Lighthouse JSON and then do the `GROUP BY`. I wish I could come up with a query but I don't think it's possible so here's my magical wand version :)

**Currently**
```sql
SELECT 
  GET_ORIGIN(url) as origin,
  /* averageExecutionTime is totalExecutionTime / popularity */
  COUNT(executionTime) as popularity,
  SUM(executionTime) as totalExecutionTime 
FROM (
  SELECT
    JSON_EXTRACT(report, '$.audits.bootup-time.details.items[0].url') AS url,
    FLOAT(JSON_EXTRACT(report, '$.audits.bootup-time.details.items[0].scripting')) as executionTime
  FROM
    [httparchive:lighthouse.2019_03_01_mobile],
  ... /* continued for all the indexes of .details.items */
)
GROUP BY origin
```

**Ideally**
```sql
SELECT 
  entity,
  /* averageExecutionTime is totalExecutionTime / popularity */
  COUNT(executionTime) as popularity,
  SUM(executionTime) as totalExecutionTime 
FROM (
  /* no clue how to write this part so we would have already done a `GROUP BY entity` on each page */
)
GROUP BY entity
```

**Comment by kdzwinel** [comment]
> we essentially need to run the require('third-party-web').getEntity function on each entry

Isn't entity list basically a `domain` → `entity name` mapping? Could this be a simple, two column, table that you could join?

**Comment by patrickhulce** [comment]
> Isn't entity list basically a domain → entity name mapping? Could this be a simple, two column, table that you could join?

That'd get us most of the way but there's fancy root domain logic that happens in there too. I suppose we could just generate a dump of all observed origins -> entity name pre-resolved

**Comment by rviscomi** [comment]
So would the aggregation be something like the median sum of execution times per page per entity?

Also FYI I added the map of domains -> entities to a BigQuery table in the HTTP Archive project: https://bigquery.cloud.google.com/table/httparchive:scratchspace.third_parties?pli=1&tab=preview. It's not kept in sync with this repo but that might be a good way to do everything in SQL.

**Comment by patrickhulce** [comment]
> So would the aggregation be something like the median sum of execution times per page per entity?

Yeah "sum of execution times per page per entity" is what we're looking for from the inner query rather than "execution time per script". Seems like adding a third layer of nesting and grouping by `page_url, entity` in the inner query would fix this. I'll have to give it a shot :)

**Comment by rviscomi** [comment]
Ah ok. I'll need to stew in this but for now here's an example of joining the entity data: https://discuss.httparchive.org/t/ads-impact-bytes-transferred-execution-time-on-the-web-over-time/1620/8?u=rviscomi

**Comment by patrickhulce** [comment]
Fixed by f6c6159a5c4a2803cf7df5bca194282b1c1dfc12 🎉

**Comment by kdzwinel** [comment]
Oh, I missed it. Great work (as always) 👏 Special thanks for mentioning most notable changes 💯
